File: array/sliding_window/fixed_size/1423_max_points.py
# There are several cards arranged in a row, and each card has an associated number of points. The points are given in the integer array cardPoints.

# In one step, you can take one card from the beginning or from the end of the row. You have to take exactly k cards.

# Your score is the sum of the points of the cards you have taken.

# Given the integer array cardPoints and the integer k, return the maximum score you can obtain.

# https://leetcode.com/problems/maximum-points-you-can-obtain-from-cards/

# construct a new array to apply sliding window
# time complexity O(k)
# space complexity O(k)
class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
        arr = cardPoints[-k:] + cardPoints[:k]
        left = right = window_sum = ans = 0

        while right < len(arr):
            window_sum += arr[right]
            if right - left + 1 == k:
                ans = max(ans, window_sum)
                window_sum -= arr[left]
                left += 1
            right += 1
        
        return ans

# minimize the cards you didn't pick: min sum of window size l - k
# time complexity O(n)
# space complexity O(1)
class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
        l, total = len(cardPoints), sum(cardPoints)
        min_sum = float('inf')
        left = right = window_sum = 0
        
        while right < l:
            window_sum += cardPoints[right]
            if right - left + 1 == l - k:
                min_sum = min(window_sum, min_sum)
                window_sum -= cardPoints[left]
                left += 1
            right += 1

        return total - min_sum if min_sum < float('inf') else total

File: array/sliding_window/fixed_size/1423_max_points.js
/*
There are several cards arranged in a row, and each card has an associated 
number of points. The points are given in the integer array cardPoints.

In one step, you can take one card from the beginning or from the end of the row. You have to take exactly k cards.

Your score is the sum of the points of the cards you have taken.

Given the integer array cardPoints and the integer k, return the maximum score 
you can obtain.

1 <= k <= cardPoints.length
*/

// sliding window -- best version: fix the overflow and slide
// time complexity O(n)
// space complexity O(1)
var maxScore = function (cardPoints, k) {
  let l = cardPoints.length;
  let left = l - k,
    right = l - k;
  let sum = 0,
    ans = 0;
  while (right < k + l) {
    sum += cardPoints[right % l];
    if (right - left + 1 >= k) {
      ans = Math.max(sum, ans);
      sum -= cardPoints[left % l];
      left++;
    }
    right++;
  }
  return ans;
};

// sliding window --  min the card didn't pick
// time complexity O(n)
// space complexity O(1)
var maxScore = function (cardPoints, k) {
  let length = cardPoints.length,
    left = 0,
    right = 0;
  let total = 0,
    sum = 0,
    ans = Infinity;
  // find the total of entire card
  for (let i = 0; i < length; i++) {
    total += cardPoints[i];
  }
  if (k == length) return total;
  // minimize the (length - k) card we didn't pick
  while (right < length) {
    sum += cardPoints[right];
    if (right - left + 1 >= length - k) {
      ans = Math.min(ans, sum);
      sum -= cardPoints[left];
      left++;
    }
    right++;
  }
  return total - ans;
};
